### PR TITLE
Hip develop event sys release

### PR DESCRIPTION
**Pull request by bensander**
Use system-scope fence where necessary, and thus re-enable multi-thread support.

### CODE REVIEW

**Comment by whchung** [comment]
Could you elaborate a bit more on the rationale to OR a release to system flag when we create an event without timing info?

**Comment by bensander** [comment]
Events have a few different usage models:
- timing commands on the GPU.  For example, enqueue start event, then kernel, then stop event.  Wait for finish and diff the stop and start times to time the kernel.
- GPU-side dependency.  For example, enqueue an event in streamA, then use hipStreamWaitEvent in streamB.
- Host-side dependency: For example, enqueue an event, then on host use hipEventSynchronize or hipEventQuery.    This case needs to use an event created with hipEventReleaseToSystem to ensure the memory is consistent.  

Basically the code here is an optimization to try to distinguish these cases, and in particular not add the fence flag when the events are used for timing.

**Comment by whchung** [comment]
I'd like to invite @parallelo to check the validity of the proposed change from application perspective.

**Comment by whchung** [comment]
@bensander should this PR go to `hip` , or `hip-develop` branch?

**Comment by bensander** [comment]
Hip-develop.
